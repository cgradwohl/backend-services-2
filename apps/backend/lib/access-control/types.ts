export type AnalyticsActions = "analytics:View";

export type ApiKeyActions =
  | "apikey:RotateKey"
  | "apikey:ListItems"
  | "apikey:ReadItem"
  | "apikey:WriteItem";

export type AuditTrailActions = "auditTrail:ListItems";

export type AutomationLogsActions =
  | "automationLogs:ListItems"
  | "automationLogs:ReadItem"
  | "automationLogs:WriteItem";

export type AutomationTemplateActions =
  | "automationTemplate:ListItems"
  | "automationTemplate:ReadItem"
  | "automationTemplate:WriteItem";

export type BillingAction =
  | "billing:UpdatePaymentMethod"
  | "billing:UpdatePlan"
  | "billing:ViewBilling";

export type BrandAction =
  | "brand:ListItems"
  | "brand:ReadItem"
  | "brand:WriteItem";

export type CategoryAction =
  | "category:ListItems"
  | "category:ReadItem"
  | "category:WriteItem";

export type IntegrationAction =
  | "integration:ListItems"
  | "integration:ReadItem"
  | "integration:WriteItem";

export type ListAction = "list:ListItems" | "list:ReadItem" | "list:WriteItem";

export type MessageAction =
  | "message:ListItems"
  | "message:ReadEventDetails"
  | "message:ReadItem"
  | "message:RequeueItem"
  | "message:WriteItem";

export type MetricsAction = "metrics:GetMetrics";

export type PreferenceTemplateAction =
  | "preferenceTemplate:ListItems"
  | "preferenceTemplate:ReadItem"
  | "preferenceTemplate:WriteItem";

export type RecipientAction =
  | "recipient:ListItems"
  | "recipient:ReadItem"
  | "recipient:WriteItem";

export type Recipient_2022_01_28_Action =
  | "recipient_2022_01_28:ListItems"
  | "recipient_2022_01_28:ReadItem"
  | "recipient_2022_01_28:WriteItem";

export type SecurityAction = "security:WriteSettings";

export type TemplateAction =
  | "template:ListItems"
  | "template:ReadItem"
  | "template:WriteItem";

export type TenantAction =
  | "tenant:ListItems"
  | "tenant:ReadItem"
  | "tenant:WriteItem";

export type TrackingAction = "tracking:WriteSettings";

export type UserAction =
  | "user:InviteUser"
  | "user:ListItems"
  | "user:ReadItem"
  | "user:WriteItem";

export type WebhookAction =
  | "webhook:ListItems"
  | "webhook:ReadItem"
  | "webhook:WriteItem";

export type WildcardAction =
  | "*"
  | "analytics:*"
  | "apikey:*"
  | "auditTrail:*"
  | "automationLogs:*"
  | "automationTemplate:*"
  | "billing:*"
  | "brand:*"
  | "category:*"
  | "integration:*"
  | "list:*"
  | "message:*"
  | "metrics:*"
  | "preferenceTemplate:*"
  | "recipient:*"
  | "recipient_2022_01_28:*"
  | "security:*"
  | "template:*"
  | "tenant:*"
  | "tracking:*"
  | "user:*"
  | "webhook:*";

export type Action =
  | AnalyticsActions
  | ApiKeyActions
  | AuditTrailActions
  | AutomationLogsActions
  | AutomationTemplateActions
  | BillingAction
  | BrandAction
  | CategoryAction
  | IntegrationAction
  | ListAction
  | MessageAction
  | MetricsAction
  | PreferenceTemplateAction
  | RecipientAction
  | Recipient_2022_01_28_Action
  | SecurityAction
  | TemplateAction
  | TenantAction
  | TrackingAction
  | UserAction
  | WebhookAction;

export type Effect = "ALLOW" | "DENY";

export interface IPolicyStatement {
  actions: Array<Action | WildcardAction>;
  effect: Effect;
  resources: string[];
}

export interface IPolicy {
  statements: IPolicyStatement[];
  version: "2020-11-09";
}

export interface IRole {
  label?: string;
  description?: string;
  key: string;
  policies: IPolicy[];
}
