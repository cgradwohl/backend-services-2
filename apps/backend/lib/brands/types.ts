import { DocumentClient } from "aws-sdk/clients/dynamodb";
import { Operation } from "fast-json-patch";
import { IPublishableObject } from "../dynamo/object-service/types";

export type BrandCourierObject = IPublishableObject<ICourierObjectJson>;

export type CreateFn = (
  tenantId: string,
  creator: string,
  brand: ICreatableBrand,
  options?: { publish?: boolean }
) => Promise<IBrand>;

export type CreateDefaultBrandFn = (
  tenantId: string,
  creator: string,
  id?: string
) => Promise<IBrand>;

export type DeleteFn = (tenantId: string, id: string) => Promise<void>;

export interface IGetFnOptions {
  extendDefaultBrand?: boolean;
}
export type GetFn = (
  tenantId: string,
  id: string,
  options?: IGetFnOptions
) => Promise<IBrand>;
export type GetDefaultFn = (tenantId: string) => Promise<IBrand>;
export type GetLatestFn = (tenantId: string, id: string) => Promise<IBrand>;
export type GetLatestDefaultFn = (tenantId: string) => Promise<IBrand>;

export type GetVersionFn = (
  tenantId: string,
  id: string,
  version: string
) => Promise<IBrand>;

export interface IBrand {
  created: number;
  creator: string;
  defaultBrand?: boolean;
  id: string;
  name: string;
  published?: number;
  settings: IBrandSettings;
  snippets?: IBrandSnippets;
  updated: number;
  updater: string;
  version: string;
}

export interface IBrandColors {
  // CSS compliant color value
  primary?: string;
  secondary?: string;
  tertiary?: string;
}

export interface IBrandSettings {
  colors?: IBrandColors;
  email?: IBrandSettingsEmail;
  inapp?: IBrandSettingsInApp;
}

interface IBrandTemplate {
  backgroundColor?: string;
  blocksBackgroundColor?: string;
  enabled: boolean;
  footer?: string;
  head?: string;
  header?: string;
  width?: string;
}

export interface IBrandTemplateOverride extends IBrandTemplate {
  mjml?: IBrandTemplate;
  footerBackgroundColor?: string;
  footerFullWidth?: boolean;
}

export interface IBrandSettingsEmail {
  templateOverride?: IBrandTemplateOverride;
  head?: {
    inheritDefault: boolean;
    content?: string;
  };
  footer?: {
    content?: object;
    inheritDefault?: boolean;
    markdown?: string;
    social?: IBrandSettingsSocialPresence;
  };
  header?: {
    inheritDefault?: boolean;
    barColor?: string;
    logo?: {
      href?: string;
      image?: string;
    };
  };
}
export interface IBrandSettingsInApp {
  borderRadius?: string;
  disableMessageIcon?: boolean;
  fontFamily?: string;
  placement?: "top" | "bottom" | "left" | "right";
  emptyState?: {
    textColor?: string;
    text?: string;
  };
  widgetBackground?: {
    topColor?: string;
    bottomColor?: string;
  };
  colors?: {
    invertHeader?: boolean;
    invertButtons?: boolean;
  };
  icons?: {
    bell?: string;
    message?: string;
  };
  preferences?: {
    templateIds: string[];
  };
}

export interface IBrandSettingsSocialPresence {
  inheritDefault?: boolean;
  facebook?: {
    url: string;
  };
  instagram?: {
    url: string;
  };
  linkedin?: {
    url: string;
  };
  medium?: {
    url: string;
  };
  twitter?: {
    url: string;
  };
}

export interface IBrandSnippet {
  format: "handlebars"; // could support other formats in the future
  name: string;
  value: string;
}
export interface IBrandSnippets {
  items: IBrandSnippet[];
}

export interface ICourierObjectJson {
  settings?: IBrandSettings;
  snippets?: IBrandSnippets;
}

export interface ICreatableBrand {
  id?: string;
  name: string;
  settings: IBrandSettings;
  snippets?: IBrandSnippets;
}
export interface IReplaceableBrand {
  name: string;
  settings: IBrandSettings;
}

export type ListFn = (
  tenantId: string,
  exclusiveStartKey?: DocumentClient.Key
) => Promise<{
  items: IBrand[];
  lastEvaluatedKey?: DocumentClient.Key;
}>;

export type ListVersionsFn = (
  tenantId: string,
  id: string
) => Promise<{
  items: IBrand[];
  lastEvaluatedKey: DocumentClient.Key;
}>;

export type PatchFn = (
  tenantId: string,
  brandId: string,
  updater: string,
  ops: Operation[]
) => Promise<IBrand>;

export type ReplaceFn = (
  tenantId: string,
  userId: string,
  id: string,
  brand: IReplaceableBrand,
  options?: { publish?: boolean }
) => Promise<IBrand>;
