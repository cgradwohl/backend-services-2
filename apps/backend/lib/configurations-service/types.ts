import {
  CourierObject,
  EventJsonWire,
  IConfigurationJson,
  INotificationJsonWire,
} from "~/types.api";
import { IGetFn } from "~/lib/dynamo/object-service/types";

type DeleteResult =
  | { status: "ok" }
  | { notifications: Array<{ id: string; title: string }>; status: "error" };

export type DeleteFn = (
  tenantId: string,
  id: string,
  userId: string
) => Promise<DeleteResult>;

export type Env = "production" | "test";
export type Arg = CourierObject<INotificationJsonWire | EventJsonWire>;
export type GetConfigurationByEnvFn = (
  configuration: CourierObject<IConfigurationJson>,
  env: Exclude<Env, "production">
) => CourierObject<IConfigurationJson>;
