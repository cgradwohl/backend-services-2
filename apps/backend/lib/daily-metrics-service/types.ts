export type MetricIncrement = string | [string, number];

export type IncrementMetrics = (
  tenantId: string,
  timestamp: Date | number,
  ...metrics: MetricIncrement[]
) => Promise<void>;

export interface IDailyMetricsKey {
  day?: string;
  tenantId: string;
}

export interface IDailyMetricsDynamo extends IDailyMetricsKey {
  day: string;
  [key: string]: any; // necessary because typescript can't separate the above types
}

export interface IDailyMetrics extends IDailyMetricsKey {
  day: string;
  metrics: {
    [key: string]: number;
  };
}
