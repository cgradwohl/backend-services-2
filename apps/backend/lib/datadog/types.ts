export type CreateMetricFn = (
  tenantId: string,
  metric: string,
  options: ICreateMetricOptions
) => IMetric;

export type HttpPostFn = (request: IHttpPostRequest) => Promise<void>;

export interface ICreateMetricOptions {
  interval?: number;
  tags?: string[];
  timestamp?: number;
  value?: number;
  type?: MetricType;
}

export interface IHttpPostRequest {
  series: Series;
}

export interface IMetric {
  metric: string;
  points: [[number, number]];
  tags?: string[];
  type?: MetricType;
}

export type IncrementFn = (
  tenantId: string,
  metric: string | string[],
  options?: {
    interval?: number;
    tags?: string[];
    timestamp?: number;
    value?: number;
  }
) => Promise<void>;

export type MetricHandlerFn = <T extends (...args: any[]) => ReturnType<T>>(
  fn: T
) => (...args: Parameters<T>) => ReturnType<T>;

type MetricType =
  | "count"
  | "distribution"
  | "gauge"
  | "histogram"
  | "rate"
  | "set";

export type Series = IMetric[];
