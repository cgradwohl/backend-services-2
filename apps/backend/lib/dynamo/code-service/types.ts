import { ICodeServiceObject } from "../../../types.internal";

export interface ICodeService<T> {
  create: (
    data: T,
    options?: {
      additionalAttributes?: { email?: string };
      transform?: (code: string) => string;
    }
  ) => Promise<ICodeServiceObject<T>>;
  queryByEmail: (email: string) => Promise<Array<ICodeServiceObject<T>>>;
  get: (code: string) => Promise<ICodeServiceObject<T>>;
  remove: (code: string) => Promise<{}>;
  queryBeginsWith: (
    beginsWith: string
  ) => Promise<Array<ICodeServiceObject<T>>>;
}
