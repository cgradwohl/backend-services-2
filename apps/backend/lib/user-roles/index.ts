import * as systemRoles from "~/lib/access-control/roles";
import { getItem, query, deleteItem, update } from "~/lib/dynamo";
import { IUserRoleService } from "./types";
import { IRole } from "../access-control/types";

const findSystemRole = (key: string): IRole => {
  for (const role in systemRoles) {
    if (systemRoles[role].key === key) {
      return systemRoles[role];
    }
  }
};

const fromDynamoItem = (item: any): IRole => {
  if (!item) {
    return;
  }

  return {
    description: item.description,
    label: item.label,
    key: item.sk.replace(/^role\//, ""),
    policies: item.policies,
  };
};

export default (tenantId: string): IUserRoleService => {
  return {
    delete: async (key) => {
      await deleteItem({
        Key: {
          pk: tenantId,
          sk: `role/${key}`,
        },
        TableName: process.env.USER_ROLES_TABLE_NAME,
      });
    },

    get: async (key) => {
      const response = await getItem({
        Key: {
          pk: tenantId,
          sk: `role/${key}`,
        },
        TableName: process.env.USER_ROLES_TABLE_NAME,
      });

      if (response?.Item) {
        return fromDynamoItem(response?.Item);
      }

      return findSystemRole(key);
    },

    list: async () => {
      const { Items: items } = await query({
        ExpressionAttributeNames: {
          "#pk": "pk",
          "#sk": "sk",
        },
        ExpressionAttributeValues: {
          ":pk": tenantId,
          ":sk": `role/`,
        },
        KeyConditionExpression: "#pk = :pk AND begins_with(#sk, :sk)",
        TableName: process.env.USER_ROLES_TABLE_NAME,
      });

      const roles: { [key: string]: IRole } = { ...systemRoles };
      // overlay custom roles in case changes are required for a system role
      for (const item of items) {
        const key = item.sk.replace(/^role\//, "");
        roles[key] = fromDynamoItem(item);
      }

      return Object.keys(roles).map((key) => ({
        ...roles[key],
      }));
    },

    replace: async (role) => {
      await update({
        ExpressionAttributeNames: {
          "#description": "description",
          "#label": "label",
          "#policies": "policies",
        },
        ExpressionAttributeValues: {
          ":description": role.description,
          ":label": role.label,
          ":policies": role.policies,
        },
        Key: {
          pk: tenantId,
          sk: `role/${role.key}`,
        },
        UpdateExpression:
          "SET #description = :description, #label = :label, #policies = :policies",
        TableName: process.env.USER_ROLES_TABLE_NAME,
      });
    },
  };
};
