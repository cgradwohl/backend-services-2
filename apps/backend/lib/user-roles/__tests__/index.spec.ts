import { IRole } from "~/lib/access-control/types";
import { getItem, query, deleteItem, update } from "~/lib/dynamo";
import service from "../index";

jest.mock("~/lib/dynamo");

const mockGetItem = getItem as jest.Mock;
const mockDeleteItem = deleteItem as jest.Mock;
const mockQuery = query as jest.Mock;
const mockUpdate = update as jest.Mock;

const mockWorkspaceId = "mockWorkspaceId";

let userRoles: ReturnType<typeof service>;

const createTestRole = (role?: Partial<IRole>): IRole => {
  return {
    label: role?.label ?? "label",
    description: role?.description ?? "description",
    key: role?.key ?? "role",
    policies: role?.policies ?? [],
  };
};

const createMockDynamoItem = (role: IRole) => ({
  pk: mockWorkspaceId,
  sk: `role/${role.key}`,
  ...role,
});

beforeEach(() => {
  userRoles = service(mockWorkspaceId);
});

afterEach(() => {
  jest.resetAllMocks();
});

describe("delete", () => {
  it("should attempt to delete item from dynamo", async () => {
    await userRoles.delete("id");
    expect(mockDeleteItem).toMatchSnapshot();
  });
});

describe("get", () => {
  it("should attempt to fetch the role from dynamo", async () => {
    await userRoles.get("custom-role");
    expect(mockGetItem).toMatchSnapshot();
  });

  describe("custom roles", () => {
    beforeEach(() => {
      const mockRole = createTestRole({ key: "custom-role" });
      mockGetItem.mockResolvedValue({
        Item: createMockDynamoItem(mockRole),
      });
    });

    it("should return a custom role when found", async () => {
      const role = await userRoles.get("custom-role");
      expect(role).toMatchSnapshot();
    });
  });

  describe("system roles", () => {
    beforeEach(() => {
      mockGetItem.mockResolvedValue({
        Item: undefined,
      });
    });

    it("should return role", async () => {
      const roleNames = [
        "ADMINISTRATOR",
        "ANALYST",
        "DESIGNER",
        "DEVELOPER",
        "MANAGER",
        "SUPPORT_SPECIALIST",
      ];
      for (const roleName of roleNames) {
        const role = await userRoles.get(roleName);
        expect(role).toMatchSnapshot();
      }
    });
  });
});

describe("list", () => {
  beforeEach(async () => {
    const mockItem = createMockDynamoItem(createTestRole());

    mockQuery.mockResolvedValue({
      Items: [mockItem],
    });
  });

  it("should query dynamo for list of items", async () => {
    await userRoles.list();
    expect(mockQuery).toMatchSnapshot();
  });

  it("should return list of items", async () => {
    const list = await userRoles.list();
    expect(list).toMatchSnapshot();
  });
});

describe("replace", () => {
  it("should attempt to replace item in dynamo", async () => {
    await userRoles.replace(createTestRole());
    expect(mockUpdate).toMatchSnapshot();
  });
});
