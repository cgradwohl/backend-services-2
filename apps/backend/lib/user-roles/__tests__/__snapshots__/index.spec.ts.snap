// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete should attempt to delete item from dynamo 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Key": Object {
          "pk": "mockWorkspaceId",
          "sk": "role/id",
        },
        "TableName": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`get custom roles should return a custom role when found 1`] = `
Object {
  "description": "description",
  "key": "custom-role",
  "label": "label",
  "policies": Array [],
}
`;

exports[`get should attempt to fetch the role from dynamo 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Key": Object {
          "pk": "mockWorkspaceId",
          "sk": "role/custom-role",
        },
        "TableName": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`get system roles should return role 1`] = `
Object {
  "description": "Best for company administrators and business owners.  Has permissions for everything.",
  "key": "ADMINISTRATOR",
  "label": "Administrator",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "*",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`get system roles should return role 2`] = `
Object {
  "description": "Best for users who need full read-only access to the platform (except logs).",
  "key": "ANALYST",
  "label": "Analyst",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "brand:ListItems",
            "brand:ReadItem",
            "category:ListItems",
            "category:ReadItem",
            "integration:ListItems",
            "integration:ReadItem",
            "list:ListItems",
            "list:ReadItem",
            "metrics:GetMetrics",
            "template:ListItems",
            "template:ReadItem",
            "user:ListItems",
            "user:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`get system roles should return role 3`] = `
Object {
  "description": "Has the ability to update templates and brands but can't update integrations or settings.",
  "key": "DESIGNER",
  "label": "Designer",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "brand:ListItems",
            "brand:ReadItem",
            "category:ListItems",
            "category:ReadItem",
            "integration:ListItems",
            "integration:ReadItem",
            "list:ListItems",
            "list:ReadItem",
            "metrics:GetMetrics",
            "template:ListItems",
            "template:ReadItem",
            "user:ListItems",
            "user:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "automationTemplate:*",
            "brand:*",
            "category:*",
            "preferenceTemplate:*",
            "template:*",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`get system roles should return role 4`] = `
Object {
  "description": "Best for engineers and developers who will primarily work with Courier's API and template designer.",
  "key": "DEVELOPER",
  "label": "Developer",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "brand:ListItems",
            "brand:ReadItem",
            "category:ListItems",
            "category:ReadItem",
            "integration:ListItems",
            "integration:ReadItem",
            "list:ListItems",
            "list:ReadItem",
            "metrics:GetMetrics",
            "template:ListItems",
            "template:ReadItem",
            "user:ListItems",
            "user:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "apikey:ReadItem",
            "apikey:ListItems",
            "automationLogs:*",
            "automationTemplate:*",
            "brand:*",
            "category:*",
            "integration:*",
            "list:*",
            "message:*",
            "preferenceTemplate:*",
            "recipient:*",
            "template:*",
            "webhook:*",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "test/*",
          ],
        },
        Object {
          "actions": Array [
            "message:ListItems",
            "message:ReadItem",
            "automationLogs:ListItems",
            "automationLogs:ReadItem",
            "automationTemplate:ListItems",
            "automationTemplate:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`get system roles should return role 5`] = `
Object {
  "description": "Best for a manager of a team that doesn't need the ability to update users or billing.",
  "key": "MANAGER",
  "label": "Manager",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "brand:ListItems",
            "brand:ReadItem",
            "category:ListItems",
            "category:ReadItem",
            "integration:ListItems",
            "integration:ReadItem",
            "list:ListItems",
            "list:ReadItem",
            "metrics:GetMetrics",
            "template:ListItems",
            "template:ReadItem",
            "user:ListItems",
            "user:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "apikey:ReadItem",
            "apikey:ListItems",
            "automationLogs:*",
            "automationTemplate:*",
            "brand:*",
            "category:*",
            "integration:*",
            "list:*",
            "message:*",
            "preferenceTemplate:*",
            "recipient:*",
            "template:*",
            "webhook:*",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`get system roles should return role 6`] = `
Object {
  "description": "Best for customer support specialists regularly use the platform but don't need to update templates or brands.",
  "key": "SUPPORT_SPECIALIST",
  "label": "Support",
  "policies": Array [
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "analytics:*",
            "brand:ListItems",
            "brand:ReadItem",
            "category:ListItems",
            "category:ReadItem",
            "integration:ListItems",
            "integration:ReadItem",
            "list:ListItems",
            "list:ReadItem",
            "metrics:GetMetrics",
            "template:ListItems",
            "template:ReadItem",
            "user:ListItems",
            "user:ReadItem",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
    Object {
      "statements": Array [
        Object {
          "actions": Array [
            "automationLogs:*",
            "list:*",
            "message:*",
            "preferenceTemplate:ListItems",
            "preferenceTemplate:ReadItem",
            "recipient:*",
          ],
          "effect": "ALLOW",
          "resources": Array [
            "*",
          ],
        },
      ],
      "version": "2020-11-09",
    },
  ],
}
`;

exports[`list should query dynamo for list of items 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ExpressionAttributeNames": Object {
          "#pk": "pk",
          "#sk": "sk",
        },
        "ExpressionAttributeValues": Object {
          ":pk": "mockWorkspaceId",
          ":sk": "role/",
        },
        "KeyConditionExpression": "#pk = :pk AND begins_with(#sk, :sk)",
        "TableName": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`list should return list of items 1`] = `
Array [
  Object {
    "description": "Best for company administrators and business owners.  Has permissions for everything.",
    "key": "ADMINISTRATOR",
    "label": "Administrator",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "*",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "Best for users who need full read-only access to the platform (except logs).",
    "key": "ANALYST",
    "label": "Analyst",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "brand:ListItems",
              "brand:ReadItem",
              "category:ListItems",
              "category:ReadItem",
              "integration:ListItems",
              "integration:ReadItem",
              "list:ListItems",
              "list:ReadItem",
              "metrics:GetMetrics",
              "template:ListItems",
              "template:ReadItem",
              "user:ListItems",
              "user:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "Has the ability to update templates and brands but can't update integrations or settings.",
    "key": "DESIGNER",
    "label": "Designer",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "brand:ListItems",
              "brand:ReadItem",
              "category:ListItems",
              "category:ReadItem",
              "integration:ListItems",
              "integration:ReadItem",
              "list:ListItems",
              "list:ReadItem",
              "metrics:GetMetrics",
              "template:ListItems",
              "template:ReadItem",
              "user:ListItems",
              "user:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "automationTemplate:*",
              "brand:*",
              "category:*",
              "preferenceTemplate:*",
              "template:*",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "Best for engineers and developers who will primarily work with Courier's API and template designer.",
    "key": "DEVELOPER",
    "label": "Developer",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "brand:ListItems",
              "brand:ReadItem",
              "category:ListItems",
              "category:ReadItem",
              "integration:ListItems",
              "integration:ReadItem",
              "list:ListItems",
              "list:ReadItem",
              "metrics:GetMetrics",
              "template:ListItems",
              "template:ReadItem",
              "user:ListItems",
              "user:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "apikey:ReadItem",
              "apikey:ListItems",
              "automationLogs:*",
              "automationTemplate:*",
              "brand:*",
              "category:*",
              "integration:*",
              "list:*",
              "message:*",
              "preferenceTemplate:*",
              "recipient:*",
              "template:*",
              "webhook:*",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "test/*",
            ],
          },
          Object {
            "actions": Array [
              "message:ListItems",
              "message:ReadItem",
              "automationLogs:ListItems",
              "automationLogs:ReadItem",
              "automationTemplate:ListItems",
              "automationTemplate:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "Best for a manager of a team that doesn't need the ability to update users or billing.",
    "key": "MANAGER",
    "label": "Manager",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "brand:ListItems",
              "brand:ReadItem",
              "category:ListItems",
              "category:ReadItem",
              "integration:ListItems",
              "integration:ReadItem",
              "list:ListItems",
              "list:ReadItem",
              "metrics:GetMetrics",
              "template:ListItems",
              "template:ReadItem",
              "user:ListItems",
              "user:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "apikey:ReadItem",
              "apikey:ListItems",
              "automationLogs:*",
              "automationTemplate:*",
              "brand:*",
              "category:*",
              "integration:*",
              "list:*",
              "message:*",
              "preferenceTemplate:*",
              "recipient:*",
              "template:*",
              "webhook:*",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "Best for customer support specialists regularly use the platform but don't need to update templates or brands.",
    "key": "SUPPORT_SPECIALIST",
    "label": "Support",
    "policies": Array [
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "analytics:*",
              "brand:ListItems",
              "brand:ReadItem",
              "category:ListItems",
              "category:ReadItem",
              "integration:ListItems",
              "integration:ReadItem",
              "list:ListItems",
              "list:ReadItem",
              "metrics:GetMetrics",
              "template:ListItems",
              "template:ReadItem",
              "user:ListItems",
              "user:ReadItem",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
      Object {
        "statements": Array [
          Object {
            "actions": Array [
              "automationLogs:*",
              "list:*",
              "message:*",
              "preferenceTemplate:ListItems",
              "preferenceTemplate:ReadItem",
              "recipient:*",
            ],
            "effect": "ALLOW",
            "resources": Array [
              "*",
            ],
          },
        ],
        "version": "2020-11-09",
      },
    ],
  },
  Object {
    "description": "description",
    "key": "role",
    "label": "label",
    "policies": Array [],
  },
]
`;

exports[`replace should attempt to replace item in dynamo 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ExpressionAttributeNames": Object {
          "#description": "description",
          "#label": "label",
          "#policies": "policies",
        },
        "ExpressionAttributeValues": Object {
          ":description": "description",
          ":label": "label",
          ":policies": Array [],
        },
        "Key": Object {
          "pk": "mockWorkspaceId",
          "sk": "role/role",
        },
        "TableName": undefined,
        "UpdateExpression": "SET #description = :description, #label = :label, #policies = :policies",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
